code/modules/mob/living/carbon/xenomorph/Castes/Queen.dm

```diff
@@ -623,52 +588,49 @@
 		flick("ovipositor_dismount_destroyed", src)
 		sleep(5)
 
-	if(ovipositor)
-		ovipositor = FALSE
-		update_icons()
-		new /obj/ovipositor(loc)
+	if(!ovipositor)
+		return
+	ovipositor = FALSE
+	update_icons()
+	new /obj/ovipositor(loc)
 
-		if(observed_xeno)
-			set_queen_overwatch(observed_xeno, TRUE)
-		zoom_out()
-
-		for(var/datum/action/A in actions)
-			qdel(A)
-
-		var/list/mobile_abilities = list(
-			/datum/action/xeno_action/xeno_resting,
-			/datum/action/xeno_action/regurgitate,
-			/datum/action/xeno_action/plant_weeds,
-			/datum/action/xeno_action/choose_resin,
-			/datum/action/xeno_action/activable/secrete_resin,
-			/datum/action/xeno_action/grow_ovipositor,
-			/datum/action/xeno_action/activable/screech,
-			/datum/action/xeno_action/activable/corrosive_acid,
-			/datum/action/xeno_action/psychic_whisper,
-		 	/datum/action/xeno_action/shift_spits,
-			/datum/action/xeno_action/activable/xeno_spit,
-			/datum/action/xeno_action/activable/larva_growth,
-			/datum/action/xeno_action/toggle_pheromones
-			)
-
-		for(var/path in mobile_abilities)
-			var/datum/action/xeno_action/A = new path()
-			A.give_action(src)
-
-
-		egg_amount = 0
-		ovipositor_cooldown = world.time + 3000 //5 minutes
-		anchored = FALSE
-		update_canmove()
-
-		if(hivenumber && hivenumber <= hive_datum.len)
-			var/datum/hive_status/hive = hive_datum[hivenumber]
-
-			for(var/mob/living/carbon/Xenomorph/L in hive.xeno_leader_list)
-				L.handle_xeno_leader_pheromones(src)
-
-		if(!instant_dismount)
-			xeno_message("<span class='xenoannounce'>The Queen has shed her ovipositor.</span>", 3, hivenumber)
+	if(observed_xeno)
+		set_queen_overwatch(observed_xeno, TRUE)
+	zoom_out()
+
+	for(var/datum/action/A in actions)
+		qdel(A)
+
+	var/list/mobile_abilities = list(
+		/datum/action/xeno_action/xeno_resting,
+		/datum/action/xeno_action/regurgitate,
+		/datum/action/xeno_action/plant_weeds,
+		/datum/action/xeno_action/choose_resin,
+		/datum/action/xeno_action/activable/secrete_resin,
+		/datum/action/xeno_action/grow_ovipositor,
+		/datum/action/xeno_action/activable/screech,
+		/datum/action/xeno_action/activable/corrosive_acid,
+		/datum/action/xeno_action/psychic_whisper,
+		/datum/action/xeno_action/shift_spits,
+		/datum/action/xeno_action/activable/xeno_spit,
+		/datum/action/xeno_action/activable/larva_growth,
+		/datum/action/xeno_action/toggle_pheromones
+		)
+
+	for(var/path in mobile_abilities)
+		var/datum/action/xeno_action/A = new path()
+		A.give_action(src)
```

Comment:
Wouldn't it be more efficient to have two lists, one for each state mobile/ovipositor and switch between them on state change ?

Reply:
yes but outside the scope, the way abilities/actions work is a separate refactor

Reply:
:+1: